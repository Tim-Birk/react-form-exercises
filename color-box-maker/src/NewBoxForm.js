import React, { useState } from 'react';

const NewBoxForm = ({ addBox }) => {
  const intialState = {
    color: '',
    width: '',
    height: '',
  };
  const [formData, setFormData] = useState(intialState);

  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setFormData((fData) => ({
      ...fData,
      [name]: value,
    }));
  };

  const handleSubmit = (evt) => {
    evt.preventDefault();
    const { color, width, height } = formData;
    addBox(color, width, height);
    setFormData(intialState);
  };

  return (
    <form onSubmit={handleSubmit}>
      <div>
        <label htmlFor='color'>Color:</label>
        <input
          id='color'
          name='color'
          value={formData.color}
          onChange={handleChange}
        />
      </div>
      <div>
        <label htmlFor='width'>Width:</label>
        <input
          type='number'
          id='width'
          name='width'
          value={formData.width}
          onChange={handleChange}
        />
        <span>px</span>
      </div>
      <div>
        <label htmlFor='height'>Height:</label>
        <input
          type='number'
          id='height'
          name='height'
          value={formData.height}
          onChange={handleChange}
        />
        <span>px</span>
      </div>
      <button>Add New</button>
    </form>
  );
};

export default NewBoxForm;
