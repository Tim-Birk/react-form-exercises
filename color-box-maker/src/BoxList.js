import React, { useState } from 'react';
import Box from './Box';
import NewBoxForm from './NewBoxForm';
import { v4 as uuid } from 'uuid';
import './BoxList.css';

const BoxList = () => {
  const [boxes, setBoxes] = useState([
    { id: uuid(), color: 'green', width: 100, height: 115 },
    { id: uuid(), color: 'yellow', width: 90, height: 85 },
    { id: uuid(), color: 'red', width: 190, height: 185 },
    { id: uuid(), color: 'black', width: 175, height: 140 },
  ]);

  const addBox = (color, width, height) => {
    const newBox = { id: uuid(), color, width, height };
    setBoxes((oldBoxes) => [...oldBoxes, newBox]);
  };

  const removeBox = (id) => {
    setBoxes(boxes.filter((box) => box.id !== id));
  };

  return (
    <>
      <h1>Box List</h1>
      <NewBoxForm addBox={addBox} />
      <div className='BoxList'>
        {boxes.map((box) => (
          <Box
            key={box.id}
            id={box.id}
            color={box.color}
            width={box.width}
            height={box.height}
            removeBox={removeBox}
          />
        ))}
      </div>
    </>
  );
};

export default BoxList;
