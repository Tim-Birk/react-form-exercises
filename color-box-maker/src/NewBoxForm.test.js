import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import NewBoxForm from './NewBoxForm';
import BoxList from './BoxList';

it('renders without crashing', function () {
  render(<NewBoxForm />);
});

it('matches snapshot', function () {
  const { asFragment } = render(<NewBoxForm />);
  expect(asFragment()).toMatchSnapshot();
});

it('can add a new box', function () {
  const { getByLabelText, queryByText, queryByTestId } = render(<BoxList />);

  // no box yet
  expect(queryByTestId('Box-purple-120-100')).not.toBeInTheDocument();

  const colorInput = getByLabelText('Color:');
  const heightInput = getByLabelText('Height:');
  const widthInput = getByLabelText('Width:');
  const submitBtn = queryByText('Add New');

  // fill out the form
  fireEvent.change(colorInput, { target: { value: 'purple' } });
  fireEvent.change(heightInput, { target: { value: 100 } });
  fireEvent.change(widthInput, { target: { value: 120 } });
  fireEvent.click(submitBtn);

  // box exists!
  expect(queryByTestId('Box-purple-120-100')).toBeInTheDocument();
  // box styled correctly
  expect(queryByTestId('Box-purple-120-100')).toHaveStyle(`
      background-color: purple;
      height: 100px;
      width: 120px;
    `);
});
