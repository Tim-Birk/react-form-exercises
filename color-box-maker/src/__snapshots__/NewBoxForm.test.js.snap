// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <div>
      <label
        for="color"
      >
        Color:
      </label>
      <input
        id="color"
        name="color"
        value=""
      />
    </div>
    <div>
      <label
        for="width"
      >
        Width:
      </label>
      <input
        id="width"
        name="width"
        type="number"
        value=""
      />
      <span>
        px
      </span>
    </div>
    <div>
      <label
        for="height"
      >
        Height:
      </label>
      <input
        id="height"
        name="height"
        type="number"
        value=""
      />
      <span>
        px
      </span>
    </div>
    <button>
      Add New
    </button>
  </form>
</DocumentFragment>
`;
